Adding Efficiency and Delta R plots

samf25 opened the pull request:
I needed Efficiency and Delta R plots for some comparisons I was working on. Adding these to main will allow anyone to be able to use them.

Commit:
fix

Commit:
bug fix

Commit:
bug fix

Commit:
TEff bounds

Commit:
theta

Commit:
Fixed efficiency plots

Commit:
fix bug

Commit:
Delta R graph -- remove theta efficiency

Commit:
Fixed graphs

kkrizka commented on src/TruthHists.cxx:
```diff
@@ -32,3 +43,33 @@ void TruthHists::fill(const EVENT::MCParticle* particle) {
+        double trackPhi = track->getPhi();
+	double dphi = phi - trackPhi;
+	double deta = eta - trackEta;
+	if (dphi > 3) { dphi -= 3.1415926535; }
```
Use [TVector2::Phi_mpi_pi](https://root.cern.ch/doc/master/classTVector2.html#ad30d7548d7a2562d86cd8d1597460a44) instead.

kkrizka commented on TrackPerf/TruthHists.hxx:
```diff
@@ -14,10 +16,12 @@ class TruthHists {
-  void fill(const EVENT::MCParticle* track);
+  void fill(const EVENT::MCParticle* particle);
+  void effi(const EVENT::MCParticle* particle, bool passed);
+  void deltaR(const EVENT::MCParticle* particle, const EVENT::Track* track);
```
Histograms correlating `MCParticle` and `Track` properties should go into `ResoHists`.

kkrizka commented on src/TrackHists.cxx:
```diff
@@ -56,3 +62,11 @@ void TrackHists::fill(const EVENT::Track* track) {
+
+void TrackHists::effi(const EVENT::Track* track, bool passed) {
+	float pt = fabs(0.3 * _Bz / track->getOmega() / 1000);
+	float eta = -std::log(std::tan((1.57079632679 - std::atan(track->getTanLambda()))/2));
```
Use [TMath::Piover2](https://root.cern.ch/doc/master/namespaceTMath.html#a705b6f0bb7c7aa1ab6f2c2e4aa2acc71).

kkrizka commented on src/TrackHists.cxx:
```diff
@@ -30,6 +30,12 @@ TrackHists::TrackHists() {
+    h_effpt = new TEfficiency("fake_pt_vs_eff", 
+      "Fake Track Rate vs pT;Reconstructed pT [GeV]; Efficiency", 50, 0, 12);
+    h_effeta = new TEfficiency("fake_eta_vs_eff", 
+      "Fake Track Rate vs Eta;Reconstructed eta; Efficiency", 50, -4, 4);
```
Make plots vs theta instead to align it as all other plots currently use that unit.

kkrizka commented on src/TrackPerfHistProc.cxx:
```diff
@@ -54,23 +54,24 @@ void TrackPerfHistProc::init() {
+  _fakeTracks = std::make_shared<TrackPerf::TrackHists>(true);
   h_number_of_fakes = new TH1F("number_of_fakes",
                                ";Number of fake tracks;Events", 100, 0, 300000);
+  h_count = new TH1F("Count","C;C",20,0,10);
```
What's this for?

samf25 replied:
Ah sorry about that, this was from debugging an issue in the Gaudi version. Removed now!

kkrizka commented:
The way that `TrackPerf` is currently designed is to write the histograms and then use a plotting package to make the necessary efficiency plots (ie: [TrackPerfPlots](https://github.com/kkrizka/TrackPerfPlots)). I am not against saving a `TEfficiency` directly, as it is indeed convenient. But it is a bit awkward to fit it into the current output structure (`real`, `fake`, `all`) which targets a specific matching of MCParticle and Track. This you can see via the need to have the `effi` argument to the histogram classes.

Maybe create a new class called `EfficiencyHists`?

Commit:
Move Efficiency to new class

samf25 commented:
Ok I've moved the efficiency plots to a new class that adds a new directory in the TFile and fixed the other issues you pointed out. Thanks for the help!

kkrizka commented on src/EfficiencyHists.cxx:
```diff
@@ -0,0 +1,34 @@
+     h_efftheta = new TEfficiency("theta_vs_eff", ";Truth theta; Efficiency", 50, -4, 4);
+  } else {
+    h_effpt = new TEfficiency("fake_pt_vs_eff",
+      "Fake Track Rate vs pT;Reconstructed pT [GeV]; Efficiency", 50, 0, 12);
```
I think you also want to change the y-axis label from "Efficiency" to "Fake Rate".

kkrizka commented on src/EfficiencyHists.cxx:
```diff
@@ -0,0 +1,34 @@
+  const double* mom = particle->getMomentum();
+  double pt = std::sqrt(std::pow(mom[0], 2) + std::pow(mom[1], 2));
+  double theta = std::atanh(mom[2] / pt);
+  if (pt > 0.5) h_efftheta->Fill(passed, theta);
```
Remove hardcoded checks on the MC particle reconstructibility to make this consistent with the other plots saved in the processor.

If you are concerned about low efficiency in theta being related to low efficiency in pT, maybe also add a 2D TEfficiency plot?

samf25 replied:
I guess that makes sense. I've always added pT cuts to position efficiency plots and vice versa in the past, but doesn't hurt anything to remove them. For right not I don't think I need the 2D plot, but if I do, I'll add it later.

Commit:
Removed Hard MCP Reconstructability checks
